File: web/src/Models/Question.ts
import { camelizeKeys } from 'humps';
import { UserVote } from ".";

export default class QuestionModel {
    _id!: string;
    content!: string;
    isAuthor!: boolean;
    hideVotes!: boolean;
    isAnonymous!: boolean;
    authorFullName: string = '';
    authorEmail: string = '';
    authorNickname: string = '';
    upvotes: number = 0;
    downvotes: number = 0;
    voted: UserVote = UserVote.None;
    read: string = 'false';
    hidden: boolean = false;

    addUserVote(voteType: UserVote) {
        if (voteType === UserVote.Up) {
            this.upvotes++;
            this.downvotes -= Number(this.voted !== UserVote.None);
            this.voted = UserVote.Up;
        } else {
            this.downvotes++;
            this.upvotes -= Number(this.voted !== UserVote.None);
            this.voted = UserVote.Down;
        }
    }

    removeUserVote(): void {
        if (this.voted === UserVote.Up) {
            this.upvotes--;
        }
        else {
            this.downvotes--;
        }
    }

    toggleRead(): void {
        this.read = this.read === 'true' ? 'false' : 'true'
    }
    
    static createQuestionFromApiResult(questionDict: any): QuestionModel {
        let question = new QuestionModel();
        for (const [key, value] of Object.entries(camelizeKeys(questionDict, function (key, convert) {
            return key === '_id' ? key : convert(key);
        }))) {
            (question as any)[key] = value
        }

        question.hideVotes = question.upvotes == null && question.read === 'false';
        return question;
    }
}

File: bridges/database/objects/question.py
from datetime import datetime
from typing import List, Optional
from dataclasses import dataclass, field
from bson.objectid import ObjectId
from bridges.database.objects.question_user_context import QuestionUserContext
from .vote import Vote
from .user import User
from .mongo_object import MongoObject


@dataclass
class Question(MongoObject):
    """
    Template of question
    """

    content: str
    author: User
    is_anonymous: Optional[bool]
    reply: Optional[str] = ""
    author_nickname: str = None
    hidden: Optional[bool] = False
    votes: List[Vote] = field(default_factory=list)
    user_contexts: List[QuestionUserContext] = field(default_factory=list)
    _id: ObjectId = None
    date: datetime = field(default_factory=datetime.now) 

    def get_api_result(self, user: User, hide_votes=False):
        upvotes = sum(1 if vote.upvote else 0 for vote in self.votes)
        downvotes = len(self.votes) - upvotes
        voted = self.get_user_vote(user)
        is_read = self.is_read(user)
        return {
            **self.as_dict(skip_id=False),
            ** {
                "upvotes": upvotes if not hide_votes or voted != 'none' or is_read == 'true' else None,
                "downvotes": downvotes if not hide_votes or voted != 'none' or is_read == 'true' else None,
                "voted": voted,
                "read": is_read,
                "isAuthor": self.author == user,
                "isAnonymous": self.is_anonymous if self.is_anonymous is not None else True,
                "authorFullName": self.author.full_name if not self.is_anonymous else None,
                "authorEmail": self.author.email if not self.is_anonymous else None,
                "author_nickname": self.author_nickname if self.is_anonymous else None
            }
        }

    def is_read(self, user: User):
        if any(ctx.read == 'true' and ctx.author == user for ctx in self.user_contexts):
            return "true"
        else:
            return "false"

    def get_user_vote(self, user: User):
        """
        How user voted on question
        """

        if any(v.upvote and v.author == user for v in self.votes):
            return 'up'
        elif any(not v.upvote and v.author == user for v in self.votes):
            return 'down'
        else:
            return 'none'
